rockcraft/oci.py

```diff
@@ -190,34 +200,44 @@ def add_layer(
         name = self.image_name.split(":", 1)[0]
         return self.__class__(image_name=f"{name}:{tag}", path=self.path)
 
-    def add_user(self, base_layer_dir: Path, tag: str, username: str, uid: int) -> None:
+    def add_user(
+        self,
+        prime_dir: Path,
+        base_layer_dir: Path,
+        username: str,
+        uid: int,
+    ) -> Path:
         """Create a new ROCK user.
 
+        :param prime_dir: Path to the user-defined parts' primed content
         :param base_layer_dir: Path to the base layer's root filesystem
-        :param tag: The ROCK's image tag
         :param username: Username to be created. Same as group name.
         :param uid: UID of the username to be created. Same as GID
         """
-        passwd_file = Path("etc/passwd")
-        group_file = Path("etc/group")
-        shadow_file = Path("etc/shadow")
-
-        existing_users = (
-            (base_layer_dir / passwd_file).read_text()
-            if (base_layer_dir / passwd_file).exists()
-            else ""
-        )
-        existing_groups = (
-            (base_layer_dir / group_file).read_text()
-            if (base_layer_dir / group_file).exists()
-            else ""
-        )
+        user_files = {"passwd": "", "group": "", "shadow": ""}
+
+        prime_dir_etc = prime_dir / "etc"
+        base_layer_dir_etc = base_layer_dir / "etc"
+        # Being cautious about possible changes (edits or removals) in
+        # /etc/{passwd,group,shadow} done by the user through overlay scripts.
+        # Basically:
```

Comment:
:+1:

Reply:
I think this merits a spread test, to ensure that we support the common case of people adding users in overlays

Reply:
sure. added in https://github.com/canonical/rockcraft/pull/233/files#diff-6adf14346ad84f5632a4ea9fbab1d93df22fbb9b16531f4261bc4a15f792cedfR23